File: go/src/deposit/src/domain/deposit_status.go
package domain

import (
	"bytes"
	"encoding/json"
	"errors"

	"github.com/aws/aws-sdk-go/service/dynamodb"
)

type DepositStatus int

// Add new states here as needed.
const (
	New DepositStatus = iota
	DepositSent
	Done
	Failed
	Returned
)

var statusToString = map[DepositStatus]string{
	New:         "NEW",
	DepositSent: "DEPOSIT_SENT",
	Done:        "DONE",
	Failed:      "FAILED",
	Returned:    "RETURNED",
}

func (d DepositStatus) String() string {
	return statusToString[d]
}

var statusToID = map[string]DepositStatus{
	"NEW":          New,
	"DEPOSIT_SENT": DepositSent,
	"DONE":         Done,
	"FAILED":       Failed,
	"RETURNED":     Returned,
}

func StringToStatus(status string) (DepositStatus, error) {
	if val, ok := statusToID[status]; ok {
		return val, nil
	}
	return 0, errors.New("invalid DepositStatus")
}

func (d DepositStatus) MarshalJSON() ([]byte, error) {
	buffer := bytes.NewBufferString(`"`)
	buffer.WriteString(statusToString[d])
	buffer.WriteString(`"`)
	return buffer.Bytes(), nil
}

func (d *DepositStatus) UnmarshalJSON(b []byte) error {
	var j string
	err := json.Unmarshal(b, &j)
	if err != nil {
		return err
	}
	val, ok := statusToID[j]
	if !ok {
		return errors.New("invalid DepositStatus: " + j)
	}
	*d = val
	return nil
}

func (d DepositStatus) MarshalDynamoDBAttributeValue(av *dynamodb.AttributeValue) error {
	value := d.String()
	av.S = &value
	return nil
}

func (d *DepositStatus) UnmarshalDynamoDBAttributeValue(av *dynamodb.AttributeValue) error {
	if av.S == nil {
		return nil
	}
	val, ok := statusToID[*av.S]
	if !ok {
		return errors.New("invalid DepositStatus: " + *av.S)
	}
	*d = val
	return nil
}

File: python/src/deposit/domain/deposit_status.py
from enum import Enum
from typing import Any


class DepositStatus(Enum):
    NEW = 0
    DEPOSIT_SENT = 1
    FAILED = 2


_STATUS_TO_STRING: dict[DepositStatus, str] = {
    DepositStatus.NEW: "NEW",
    DepositStatus.DEPOSIT_SENT: "DEPOSIT_SENT",
    DepositStatus.FAILED: "FAILED",
}

_STRING_TO_STATUS: dict[str, DepositStatus] = {
    "NEW": DepositStatus.NEW,
    "DEPOSIT_SENT": DepositStatus.DEPOSIT_SENT,
    "FAILED": DepositStatus.FAILED,
}


def deposit_status_to_string(status: DepositStatus) -> str:
    return _STATUS_TO_STRING[status]


def deposit_status_from_string(value: str) -> DepositStatus:
    try:
        return _STRING_TO_STATUS[value]
    except KeyError as exc:
        raise ValueError(f"invalid DepositStatus: {value}") from exc


def deposit_status_to_ddb(status: DepositStatus) -> dict[str, Any]:
    return {"S": _STATUS_TO_STRING[status]}


def deposit_status_from_ddb(attr: dict[str, Any]) -> DepositStatus:
    value = attr.get("S")
    if value is None:
        raise ValueError("missing DepositStatus attribute")
    return deposit_status_from_string(value)
